packages/app/obojobo-document-engine/src/scripts/common/util/inject-katex-if-needed.js

```diff
@@ -0,0 +1,45 @@
+import insertDomTag from './insert-dom-tag'
+
+const injectKatexIfNeeded = async ({ value: draftModel }) => {
+	// DETECT LATEX USAGE
+	// LATEX can be in seen in 2 places
+	// HTML allows div with a classname of 'latex' to render latex
+	// AND in text nodes with a styleList item of type _latex
+	// The second can be matched against `"type":"_latex"` without many false positives
+	// However, the html node's classname will generate more false positives
+	const stringModel = JSON.stringify(draftModel)
+
+	if (stringModel.includes('latex')) {
+		const jsProps = {
+			src: 'https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.js'
+		}
+		const cssProps = {
+			rel: 'stylesheet',
+			type: 'text/css',
+			href: 'https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css'
+		}
+		insertDomTag(jsProps, 'script')
+		insertDomTag(cssProps, 'link')
+
+		await new Promise((resolve, reject) => {
+			const timeout = 15000
+			const intervalTime = 20
+			let timer = 0
+			const latexCheckInterval = setInterval(() => {
+				timer += intervalTime
+				if (window.katex) {
+					clearInterval(latexCheckInterval)
+					return resolve()
+				}
+				if (timer > timeout) {
+					clearInterval(latexCheckInterval)
+					return reject('Timed out loading katex library')
+				}
+			}, intervalTime)
+		})
+	}
+
+	return draftModel
+}
```

Comment:
I played around with an idea that would be able to eliminate false positives:

```javascript
import insertDomTag from './insert-dom-tag'

const injectKatexIfNeeded = async ({ value: draftModel }) => {
	if (usesLatex(draftModel)) {
		await injectKatex()
	}

	return draftModel
}

// Returns true if node contains a textGroup which contains an item featuring a "latex" styleRange
const hasLatexStyleRange = node => {
	// Abort if node does not contain a textGroup
	if (!node.content || !node.content.textGroup) {
		return false
	}

	// Store the textGroup into 'tg' (Tables contains their actual textGroup at textGroup.textGroup)
	let tg
	if (node.content.textGroup.textGroup) {
		tg = node.content.textGroup.textGroup
	} else {
		tg = node.content.textGroup
	}

	// If the textGroup item isn't an array then abort
	if (!Array.isArray(tg)) {
		return false
	}

	// Search every textGroupItem...
	return tg.some(textGroupItem => {
		// Abort if the textGroupItem does not contain an array at text.styleList
		if (
			!textGroupItem.text ||
			!textGroupItem.text.styleList ||
			!Array.isArray(textGroupItem.text.styleList)
		) {
			return false
		}

		// Return true if one of the textGroupItems contain a styleRange of type "_latex"
		return textGroupItem.text.styleList.some(styleRange => {
			return styleRange.type && styleRange.type === '_latex'
		})
	})
}

// Returns true of node is an HTML node containing any element with class="latex"
const hasHTMLLatex = node => {
	if (node.type !== 'ObojoboDraft.Chunks.HTML') {
		return false
	}

	const domElement = document.createElement('div')
	domElement.innerHTML = node.content.html

	return domElement.querySelector('.latex')
}

// Returns true if node is a MathEquation, an HTML node with a "latex" class or a node with a
// textGroup containing a "_latex" styleRange
const usesLatex = node => {
	if (
		node.type === 'ObojoboDraft.Chunks.MathEquation' ||
		hasLatexStyleRange(node) ||
		hasHTMLLatex(node)
	) {
		return true
	}

	if (!node.children) {
		return false
	}

	return node.children.some(usesLatex)
}

const injectKatex = async () => {
	const jsProps = {
		src: 'https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.js'
	}
	const cssProps = {
		rel: 'stylesheet',
		type: 'text/css',
		href: 'https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css'
	}
	insertDomTag(jsProps, 'script')
	insertDomTag(cssProps, 'link')

	await new Promise((resolve, reject) => {
		const timeout = 15000
		const intervalTime = 20
		let timer = 0
		const latexCheckInterval = setInterval(() => {
			timer += intervalTime
			if (window.katex) {
				clearInterval(latexCheckInterval)
				return resolve()
			}
			if (timer > timeout) {
				clearInterval(latexCheckInterval)
				return reject('Timed out loading katex library')
			}
		}, intervalTime)
	})
}

export default injectKatexIfNeeded
```

I performance tested this code on a few non-trivial and trivial modules. It's slower than the JSON approach but still is under 1ms. Typically the JSON code was running at 0.05 to 0.1ms, and this code runs at around 0.25ms to 0.8ms.

Reply:
This is cool.  Def trades accuracy for complexity though.  Other options I had in mind:

* do this check on the server - on save (and store in metadata)
* do the latex string search before the api parses the json (right now we convert text to json, which we need, but also convert it back to text to do a simple scan for a matching string)
* register functions in each node that let it scan the document and reply with scripts that need to be loaded